### ISSUE TITLE
PersistentVolumeClaims (and others) are not necessarily applied first

### ISSUE THREAD

**Issue by landism** [windmilleng/tilt#2348]
When we have a bunch of k8s yaml, some objects depend on other objects, and if A depends on B, we need ensure B is applied first (e.g., if A is in namespace B, or A has PersistentVolumeClaim B). Otherwise, A's application may fail because B does not exist.

Tilt currently attempts to do this by putting namespaces and CRDs first, but misses other cases such as PersistentVolumeClaims ([code](https://github.com/windmilleng/tilt/blob/92a183dcf837504191e851c18ec8a7fd0cd7422f/internal/k8s/entity.go#L205)).

Kustomize's approach is [much more thorough](https://github.com/kubernetes-sigs/kustomize/blob/b878cd050d5ffe5d485b34b496c915fa96405db9/pkg/gvk/gvk.go#L82).

Presumably a simple type ordering solves the vast majority of these problems, since it seems fine for Kustomize. I don't know if there are real or theoretical cases where, e.g. an object of type A could depend on an object of type B and also vice versa, at which point we'd need to get into which specific objects depend on which specific objects. 

IIUC, `kubectl apply` solves this problem, but I don't know how. Any solution it has would only work if all the objects are provided in a single invocation, and thus Tilt doesn't get it for free, since Tilt breaks application up into multiple invocations of `kubectl apply`. I looked briefly for any code that does this in k8s and didn't find it.

**Comment by jazzdan** [comment]
>Tilt currently attempts to do this by putting namespaces and CRDs first, but misses other cases such as CRDs 

@landism did you mean to mention CRDs twice here?

**Comment by landism** [comment]
> @landism did you mean to mention CRDs twice here?

fixed

**Comment by maiamcc** [comment]
> IIUC, kubectl apply solves this problem, but I don't know how
I'm not seeing that, fwiw. I'm playing with the examples in the [Persistent Volume Storage tutorial](https://kubernetes.io/docs/tasks/configure-pod-container/configure-persistent-volume-storage/), and it seems (according to `kubectl apply --dry-run` and to observed behavior) that `kubectl apply` respects the order in the yaml.

e.g. if I apply a yaml file containing:
```
<volume claim yaml>
---
<volume yaml>
```

the volume claim comes up in not too much time, but when i `describe` it I see a previous`ProvisioningFailed` error: `storageclass.storage.k8s.io "manual" not found` (implying that it errored while trying to spin up before the Volume existed).

When i apply a YAML containing the definitions in the opposite order--
```
<volume yaml>
---
<volume claim yaml>
```
i see no such error in the volume claim.

tldr: i DON'T actually think `kubectl apply` controls for this (tho I could be wrong)--but any issues that arise b/c resources are created in the wrong order seem to resolve themselves pretty quickly?

**Comment by maiamcc** [comment]
So, we can't easily control the order of yaml entities between Tilt resources: we apply the unresourced yaml first on the assumption that the yaml attached to resources has things like pods and deployments that aren't prereqs for any other yaml we might apply.

As far as the unresourced yaml though, we we CAN control the order.

[repro case here](https://github.com/windmilleng/yaml_ordering): if we let tilt apply UnresourcedYaml[PVC, PV] and then Resource[Pod], the pod is pending for a whiiile; the pod has to wait for the PVC, but the PVC has to wait for the PV.
![Screen Shot 2019-10-16 at 2 42 34 PM](https://user-images.githubusercontent.com/6332648/66948813-65907b80-f023-11e9-8b54-93f8b6b4c4bb.png)

A quick test verifies that if we apply UnresourcedYaml[PV, PVC] and then Resource[Pod], the pod goes green right away.

I'll start looking into how to sort the UnresourcedYaml in a useful way -- the snag is going to be that we handle an entity differently based on whether it's mutable or not.
